sound/soc/sof/topology.c

```diff
@@ -146,11 +148,20 @@ static int sof_keyword_dapm_event(struct snd_soc_dapm_widget *w,
 	dev_dbg(sdev->dev, "received event %d for widget %s\n",
 		event, w->name);
 
+	/* get runtime PCM params using widget's stream name */
+	spcm = snd_sof_find_spcm_name(sdev, swidget->widget->sname);
+
 	/* process events */
 	switch (event) {
 	case SND_SOC_DAPM_PRE_PMU:
+		if (spcm->stream[stream].d0i3_active) {
+			dev_dbg(sdev->dev, "KWD worked in suspended, no need to start it\n");
+			spcm->stream[stream].d0i3_active = false;
```

Comment:
why would you turn it off here?

Reply:
Maybe I should add explain about how this work in the commit message:

1. PCM trigger suspend (set d0i3_active) -> 
2. DAPM event handler for POST_PMD (d0i3_active set, ignore the event) --> 
3. DAPM event handler for PRE_PMU (d0i3_active set, no need to handle this power up event) -> 
4. PCM trigger resume (don't do real resuming as d0i3_active already set, clear d0i3_active after done).

I originally put this turning off in step 4 as the steps stated above, but then I realize this PCM trigger resume will never happen with our driver, that's why move it to the earlier step 3. 

Hi @ranj063 do you know why pcm trigger resume will never happen with our driver, I recall you did something related (suspend_on_resume?)?

Reply:
@keyonjie The logic seems fine. Setting/clearing d0i3_active seems a bit complicated, but I understand we need some mechanism to distribute this information to widget events. I'd be also ok with "d0i3_ignore_dapm_events" that is set unless we are going to S3.